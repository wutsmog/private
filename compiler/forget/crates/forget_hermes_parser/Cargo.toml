[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }