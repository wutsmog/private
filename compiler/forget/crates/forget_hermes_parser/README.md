# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.