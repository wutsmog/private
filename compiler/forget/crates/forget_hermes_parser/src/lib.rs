mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
