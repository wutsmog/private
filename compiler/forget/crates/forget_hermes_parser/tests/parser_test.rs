use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
