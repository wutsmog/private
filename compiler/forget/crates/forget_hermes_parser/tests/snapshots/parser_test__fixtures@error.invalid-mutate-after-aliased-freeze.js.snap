---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/error.invalid-mutate-after-aliased-freeze.js
---
Input:
function Component(props) {
  let x = [];
  let y = x;

  if (props.p1) {
    x = [];
  }

  let _ = <Component x={x} />;

  // y is MaybeFrozen at this point, since it may alias to x
  // (which is the above line freezes)
  y.push(props.p2);

  return <Component x={x} y={y} />;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 9,
          "end": 18
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 19,
            "end": 24
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "let",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 34,
                    "end": 35
                  }
                },
                "init": {
                  "type": "ArrayExpression",
                  "elements": [],
                  "loc": null,
                  "range": {
                    "start": 38,
                    "end": 40
                  }
                },
                "loc": null,
                "range": {
                  "start": 34,
                  "end": 40
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 30,
              "end": 41
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "let",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 48,
                    "end": 49
                  }
                },
                "init": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 52,
                    "end": 53
                  }
                },
                "loc": null,
                "range": {
                  "start": 48,
                  "end": 53
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 44,
              "end": 54
            }
          },
          {
            "type": "IfStatement",
            "test": {
              "type": "MemberExpression",
              "object": {
                "type": "Identifier",
                "name": "props",
                "typeAnnotation": null,
                "loc": null,
                "range": {
                  "start": 62,
                  "end": 67
                }
              },
              "property": {
                "type": "Identifier",
                "name": "p1",
                "typeAnnotation": null,
                "loc": null,
                "range": {
                  "start": 68,
                  "end": 70
                }
              },
              "computed": false,
              "loc": null,
              "range": {
                "start": 62,
                "end": 70
              }
            },
            "consequent": {
              "type": "BlockStatement",
              "body": [
                {
                  "type": "ExpressionStatement",
                  "expression": {
                    "type": "AssignmentExpression",
                    "operator": "=",
                    "left": {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 78,
                        "end": 79
                      }
                    },
                    "right": {
                      "type": "ArrayExpression",
                      "elements": [],
                      "loc": null,
                      "range": {
                        "start": 82,
                        "end": 84
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 78,
                      "end": 84
                    }
                  },
                  "directive": null,
                  "loc": null,
                  "range": {
                    "start": 78,
                    "end": 85
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 72,
                "end": 89
              }
            },
            "alternate": null,
            "loc": null,
            "range": {
              "start": 58,
              "end": 89
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "let",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "_",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 97,
                    "end": 98
                  }
                },
                "init": {
                  "type": "JSXElement",
                  "openingElement": {
                    "type": "JSXOpeningElement",
                    "name": {
                      "type": "JSXIdentifier",
                      "name": "Component",
                      "loc": null,
                      "range": {
                        "start": 102,
                        "end": 111
                      }
                    },
                    "attributes": [
                      {
                        "type": "JSXAttribute",
                        "name": {
                          "type": "JSXIdentifier",
                          "name": "x",
                          "loc": null,
                          "range": {
                            "start": 112,
                            "end": 113
                          }
                        },
                        "value": {
                          "type": "JSXExpressionContainer",
                          "expression": {
                            "type": "Identifier",
                            "name": "x",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": {
                              "start": 115,
                              "end": 116
                            }
                          },
                          "loc": null,
                          "range": {
                            "start": 114,
                            "end": 117
                          }
                        },
                        "loc": null,
                        "range": {
                          "start": 112,
                          "end": 117
                        }
                      }
                    ],
                    "selfClosing": true,
                    "loc": null,
                    "range": {
                      "start": 101,
                      "end": 120
                    }
                  },
                  "children": [],
                  "closingElement": null,
                  "loc": null,
                  "range": {
                    "start": 101,
                    "end": 120
                  }
                },
                "loc": null,
                "range": {
                  "start": 97,
                  "end": 120
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 93,
              "end": 121
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 225,
                    "end": 226
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "push",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 227,
                    "end": 231
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 225,
                  "end": 231
                }
              },
              "arguments": [
                {
                  "type": "MemberExpression",
                  "object": {
                    "type": "Identifier",
                    "name": "props",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 232,
                      "end": 237
                    }
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "p2",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 238,
                      "end": 240
                    }
                  },
                  "computed": false,
                  "loc": null,
                  "range": {
                    "start": 232,
                    "end": 240
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 225,
                "end": 241
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 225,
              "end": 242
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "JSXElement",
              "openingElement": {
                "type": "JSXOpeningElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "Component",
                  "loc": null,
                  "range": {
                    "start": 254,
                    "end": 263
                  }
                },
                "attributes": [
                  {
                    "type": "JSXAttribute",
                    "name": {
                      "type": "JSXIdentifier",
                      "name": "x",
                      "loc": null,
                      "range": {
                        "start": 264,
                        "end": 265
                      }
                    },
                    "value": {
                      "type": "JSXExpressionContainer",
                      "expression": {
                        "type": "Identifier",
                        "name": "x",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 267,
                          "end": 268
                        }
                      },
                      "loc": null,
                      "range": {
                        "start": 266,
                        "end": 269
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 264,
                      "end": 269
                    }
                  },
                  {
                    "type": "JSXAttribute",
                    "name": {
                      "type": "JSXIdentifier",
                      "name": "y",
                      "loc": null,
                      "range": {
                        "start": 270,
                        "end": 271
                      }
                    },
                    "value": {
                      "type": "JSXExpressionContainer",
                      "expression": {
                        "type": "Identifier",
                        "name": "y",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 273,
                          "end": 274
                        }
                      },
                      "loc": null,
                      "range": {
                        "start": 272,
                        "end": 275
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 270,
                      "end": 275
                    }
                  }
                ],
                "selfClosing": true,
                "loc": null,
                "range": {
                  "start": 253,
                  "end": 278
                }
              },
              "children": [],
              "closingElement": null,
              "loc": null,
              "range": {
                "start": 253,
                "end": 278
              }
            },
            "loc": null,
            "range": {
              "start": 246,
              "end": 279
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 26,
          "end": 281
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 0,
        "end": 281
      },
      "loc": null,
      "range": {
        "start": 0,
        "end": 281
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 281
  }
}
